Stratis.Bitcoin.Features.Notifications/BlockNotification.cs

```diff
@@ -76,57 +78,73 @@ public virtual void SyncFrom(uint256 startHash)
+        {
+            // Not syncing until the StartHash has been set.
+            if (this.StartHash == null)
+                return Task.CompletedTask;
+
+            // Not syncing until the chain is downloaded at least up to this block.
+            ChainedBlock startBlock = this.Chain.GetBlock(this.StartHash);
+            if (startBlock == null)
+                return Task.CompletedTask;
 
-                this.logger.LogTrace("Puller location set to block: {0}.", previousBlock);
+            // Sets the location of the puller to the block preceding the one we want to receive.
+            ChainedBlock previousBlock = this.Chain.GetBlock(startBlock.Height > 0 ? startBlock.Height - 1 : 0);
+            this.Puller.SetLocation(previousBlock);
+            this.tip = previousBlock;
 
-                // Send notifications for all the following blocks.
-                while (!this.reSync)
+            this.logger.LogTrace("Puller location set to block: {0}.", previousBlock);
+
+            // Send notifications for all the following blocks.
+            while (!this.reSync)
+            {
+                token.ThrowIfCancellationRequested();
+
+                var block = this.Puller.NextBlock(token);
+                if (block != null)
                 {
-                    var block = this.Puller.NextBlock(token);
-                    if (block != null)
-                    {
-                        // Broadcast the block to the registered observers.
-                        this.signals.SignalBlock(block);
-                        this.tip = this.Chain.GetBlock(block.GetHash());
-                    }
-                    else
-                    {
-                        // In reorg we reset the puller to the fork.
-                        // When a reorg happens the puller is pushed back and continues from the current fork.
-                        // Find the location of the fork.
-                        while (this.Chain.GetBlock(this.tip.HashBlock) == null)
-                            this.tip = this.tip.Previous;
-
-                        // Set the puller to the fork location.
-                        this.Puller.SetLocation(this.tip);
-                    }
+                    // Broadcast the block to the registered observers.
+                    this.signals.SignalBlock(block);
+                    this.tip = this.Chain.GetBlock(block.GetHash());
+
+                    continue;
                 }
 
-                this.reSync = false;
+                // In reorg we reset the puller to the fork.
+                // When a reorg happens the puller is pushed back and continues from the current fork.
+                // Find the location of the fork.
+                while (this.Chain.GetBlock(this.tip.HashBlock) == null)
+                    this.tip = this.tip.Previous;
+
+                // Set the puller to the fork location.
+                this.Puller.SetLocation(this.tip);
```

Comment:
The  first return `Task.CompletedTask` will break out of the loop right?
This is different form the previous logic, perhaps just remove the `  return Task.CompletedTask;`

Reply:
Ah yes, the "While" will execute until the SyncFrom method sets the reSync to true.